packages/bot/src/scratch/blocks/Text/text_getSubstring.js

```diff
@@ -138,47 +138,66 @@ Blockly.JavaScript.text_getSubstring = block => {
         text.match(/^'?\w+'?$/) ||
         (where1 !== 'FROM_END' && where1 !== 'LAST' && where2 !== 'FROM_END' && where2 !== 'LAST')
     ) {
-        if (where1 === 'FROM_START') {
-            at1 = Blockly.JavaScript.getAdjusted(block, 'AT1');
-        } else if (where1 === 'FROM_END') {
-            at1 = Blockly.JavaScript.getAdjusted(block, 'AT1', 1, false, Blockly.JavaScript.ORDER_SUBTRACTION);
-            at1 = `${text}.length - ${at1}`;
-        } else if (where1 === 'FIRST') {
-            at1 = '0';
+        switch (where1) {
+            case ('FROM_START'): {
+                at1 = Blockly.JavaScript.getAdjusted(block, 'AT1');
+                break;
+            }
+            case ('FROM_END'): {
+                at1 = Blockly.JavaScript.getAdjusted(block, 'AT1', 1, false, Blockly.JavaScript.ORDER_SUBTRACTION);
+                at1 = `${text}.length - ${at1}`;
+                break;
+            }
+            case ('FIRST'): {
+                at1 = '0';
+                break;
+            }
+            default:
+                break;
         }
 
-        if (where2 === 'FROM_START') {
-            at2 = Blockly.JavaScript.getAdjusted(block, 'AT2');
-        } else if (where2 === 'FROM_END') {
-            at2 = Blockly.JavaScript.getAdjusted(block, 'AT2', 0, false, Blockly.JavaScript.ORDER_SUBTRACTION);
-            at2 = `${text}.length - ${at2}`;
-        } else if (where2 === 'LAST') {
-            at2 = `${text}.length`;
+        switch (where2) {
+            case ('FROM_START'): {
+                at2 = Blockly.JavaScript.getAdjusted(block, 'AT2', 1);
+                break;
+            }
+            case ('FROM_END'): {
+                at2 = Blockly.JavaScript.getAdjusted(block, 'AT2', 0, false, Blockly.JavaScript.ORDER_SUBTRACTION);
+                at2 = `${text}.length - ${at2}`;
+                break;
+            }
+            case ('LAST'): {
+                at2 = `${text}.length`;
+                break;
+            }
+            default:
+                break;
         }
+
+        code = `${text}.slice(${at1}, ${at2})`;
     } else {
         at1 = Blockly.JavaScript.getAdjusted(block, 'AT1');
         at2 = Blockly.JavaScript.getAdjusted(block, 'AT2');
 
-        // binary-bot: Blockly.JavaScript.text.getIndex_ (Blockly)
-        const getIndex = (stringName, where, optAt) => {
+        const getIndex = (string_name, where, opt_at) => {
             if (where === 'FIRST') {
                 return '0';
             } else if (where === 'FROM_END') {
-                return `${stringName}.length - 1 - ${optAt}`;
+                return `${string_name}.length - 1 - ${opt_at}`;
             } else if (where === 'LAST') {
-                return `${stringName}.length - 1`;
+                return `${string_name}.length - 1`;
             }
-            return optAt;
+            return opt_at;
         };
-        const wherePascalCase = {
+        const where_pascal_case = {
```

Comment:
@aaron-binary you have the constant, why not using it?

Reply:
- We use this constant to generate a unique function name, see from line 200, also this object doesn't match the values used in the switch-case statement above so it wouldn't be useful there.